engines/wintermute/base/gfx/opengl/x/meshx.cpp

```diff
@@ -0,0 +1,486 @@
+/* ResidualVM - A 3D game interpreter
+ *
+ * ResidualVM is the legal property of its developers, whose names
+ * are too numerous to list here. Please refer to the COPYRIGHT
+ * file distributed with this source distribution.
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License
+ * as published by the Free Software Foundation; either version 2
+ * of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.	 See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
+ *
+ */
+
+/*
+ * This file is based on WME.
+ * http://dead-code.org/redir.php?target=wme
+ * Copyright (c) 2003-2013 Jan Nedoma and contributors
+ */
+
+#include "common/math.h"
+#include "engines/wintermute/base/base_game.h"
+#include "engines/wintermute/base/gfx/opengl/base_surface_opengl3d.h"
+#include "engines/wintermute/base/gfx/opengl/material.h"
+#include "engines/wintermute/base/gfx/opengl/shadow_volume.h"
+#include "engines/wintermute/base/gfx/opengl/x/frame_node.h"
+#include "engines/wintermute/base/gfx/opengl/x/meshx.h"
+#include "engines/wintermute/base/gfx/opengl/x/modelx.h"
+#include "engines/wintermute/base/gfx/opengl/x/loader_x.h"
+#include "engines/wintermute/dcgf.h"
+
+namespace Wintermute {
+
+//////////////////////////////////////////////////////////////////////////
+MeshX::MeshX(BaseGame *inGame) : BaseNamedObject(inGame) {
+	_numAttrs = 0;
+	_maxFaceInfluence = 0;
+
+	_adjacency = nullptr;
+	_skinAdjacency = nullptr;
+
+	_BBoxStart = _BBoxEnd = Math::Vector3d(0.0f, 0.0f, 0.0f);
+	_skinnedMesh = true;
+}
+
+//////////////////////////////////////////////////////////////////////////
+MeshX::~MeshX() {
+	delete[] _adjacency;
+	delete[] _skinAdjacency;
+	delete[] _vertexPositionData;
+
+	for (uint32 i = 0; i < _materials.size(); i++) {
+		delete _materials[i];
+	}
+
+	_materials.clear();
+}
+
+//////////////////////////////////////////////////////////////////////////
+bool MeshX::loadFromX(const Common::String &filename, XFileLexer &lexer) {
+	bool res = true;
+
+	lexer.advanceOnOpenBraces();

```

Comment:
Is "advanceOnOpenBraces" viable for the binary format?

Reply:
Should be, it just checks whether the current token is an open braces and advances if so. This handles optional object names. In the case above though, I forgot a lexer.next to skip the name in case there is one